dockstore-integration-testing/src/test/java/io/dockstore/common/CommonTestUtilities.java

```diff
@@ -358,13 +359,83 @@ public static List<String> listMigrations(String... additionals) {
         String configPath) {
+        String migrationsJoined = String.join(",", migrations);
         try {
-            application.run("db", "migrate", configPath, "--include", String.join(",", migrations));
+            if (!restoreMigratedDb(migrationsJoined)) {
+                application.run("db", "migrate", configPath, "--include", migrationsJoined);
+                dumpMigratedDb(migrationsJoined);
+            }
         } catch (Exception e) {
             fail("database migration failed");
         }
     }
 
+    private static String pathOfMigratedDb(String migrationsId) {
+        return "/tmp/dockstore_dump_" + migrationsId + ".sql";
+    }
+
+    private static boolean dumpMigratedDb(String migrationsId) {
+        if (!shouldCacheMigrations()) {
+            return false;
+        }
+        String path = pathOfMigratedDb(migrationsId);
+        boolean success = runCommand(dockerizeIfNecessary(String.format("pg_dump webservice_test -U postgres > %s", path)));
+        if (!success) {
+            LOG.error("dump failed");
+            runCommand(String.format("rm -f %s", path));
+        }
+        return success;
+    }
+
+    private static boolean restoreMigratedDb(String migrationsId) {
+        if (!shouldCacheMigrations()) {
+            return false;
+        }
+        String path = pathOfMigratedDb(migrationsId);
+        if (!new File(path).exists()) {
+            LOG.info("no dump exists");
+            return false;
+        }
+        boolean success = runCommand(dockerizeIfNecessary(String.format("psql webservice_test -U postgres < %s", path)));
+        if (success) {
+            runCommand(String.format("echo %s >> /tmp/used_dumps.txt", path));
+        }
+        return success;
+    }
+
+    private static boolean getEnvBoolean(String name) {
+        return Boolean.parseBoolean(System.getenv(name));
+    }
+
+    private static String dockerizeIfNecessary(String command) {
+        if (getEnvBoolean("DOCKSTORE_DOCKER_DB")) {
+            return "docker exec -i postgres1 " + command;
+        } else {
+            return command;
+        }
+    }
+
+    private static boolean shouldCacheMigrations() {
+        return getEnvBoolean("DOCKSTORE_CACHE_MIGRATIONS");
+    }
+
+    private static boolean runCommand(String command) {
+        LOG.info("running command: " + command);
+        try {
+            Process process = Runtime.getRuntime().exec(new String[]{"/bin/sh", "-c", command});
+            boolean success = process.waitFor() == 0;
+            if (!success) {
+                LOG.error("command '" + command + "' failed");
+                LOG.error("stderr: " + IOUtils.toString(process.getErrorStream(), StandardCharsets.UTF_8));
+            }
+            return success;
+        } catch (Exception e) {
+            String message = "exception running command '" + command + "'";
+            LOG.error(message);
```

Comment:
You never know, might be good to log the exception proper, although maybe it effectively gets logged with the throw on line 435?